### PR TITLE
feat: support sasl_server_init/new/step nif function

### PR DESCRIPTION
**Pull request by zhongwencool**

### CODE REVIEW

File: c_src/sasl_auth.c [hunk 1]
```diff
@@ -36,7 +36,8 @@ typedef struct {
 } sasl_state_t;
 
 // SASL connection state
```
Comment (kjellwinblad) [inline c_src/sasl_auth.c]: Please describe what the types sasl_cli_resource and sasl_srv_resource are used for in a comment. I think the current comment "SASL connection state" is confusing. These variables are used to represent types and they don't hold any connection state. Please also describe what the differences between sasl_cli_resource and sasl_srv_resource are. After reading much of the changes code I understand that cli stands for client and srv stands for server but this was far from obvious initially. Maybe these variables could be renamed to

`sasl_client_connection_state_nif_resource_type` and
`sasl_server_connection_state_nif_resource_type`

to make it easier to understand what these variables represent?

File: c_src/sasl_auth.c [hunk 2]
```diff
@@ -118,14 +120,15 @@ static ERL_NIF_TERM sasl_cli_new(ErlNifEnv* env, int UNUSED(argc), const ERL_NIF
...
     if ((!enif_inspect_binary(env, argv[0], &service))
         || (!enif_inspect_binary(env, argv[1], &host))
         || (!enif_inspect_binary(env, argv[2], &principal))) {
         return enif_make_badarg(env);
     }
 
-    state = enif_alloc_resource(sasl_resource, sizeof(sasl_state_t));
+    state = enif_alloc_resource(sasl_cli_resource, sizeof(sasl_state_t));
 
     if (!state) {
```
Comment (kjellwinblad) [inline c_src/sasl_auth.c:235]: Not introduced by this PR but this looks strange to me. If state is NULL (0), shall we really call `enif_release_resource(NULL)`?

File: c_src/sasl_auth.c [hunk 3]
```diff
@@ -141,7 +144,6 @@ static ERL_NIF_TERM sasl_cli_new(ErlNifEnv* env, int UNUSED(argc), const ERL_NIF
 
     memcpy(state->callbacks, callbacks, sizeof(callbacks));
 
-    int result;
```
Comment (kjellwinblad) [inline c_src/sasl_auth.c:268]: I guess this is a matter of style but I think it is better to declare the variable as close as possible to where it is used. You could write:

`int result = sasl_client_new((const char*)service.data, (const char*)host.data, NULL, NULL,
        state->callbacks, 0, &state->conn);`

This way it is easier to understand why the variable exists.

File: c_src/sasl_auth.c [hunk 4]
```diff
@@ -243,6 +245,116 @@ static ERL_NIF_TERM sasl_cli_step(ErlNifEnv* env, int UNUSED(argc), const ERL_NI
     return ret;
 }
 
+// server begin
+static ERL_NIF_TERM sasl_srv_new(ErlNifEnv* env, int UNUSED(argc), const ERL_NIF_TERM argv[])
+{
+    ErlNifBinary service, principal;
+    sasl_state_t* state = NULL;
```
Comment (kjellwinblad) [inline c_src/sasl_auth.c]: nit: Setting state to NULL seems unnecessary here. I would prefer if state and result are declared where they are initialized (C99 and later supports this).

**Review by kjellwinblad — COMMENTED** [review]
Are the new functions going to be used to provide Kerberos authentication for clients connecting to EMQX? 

As far as I can see the code looks correct. I just have a few minor comments about style.

I think it would be great if you could add tests for the new functions in `sasl_auth_SUITE.erl`. As we new provide functions both for the server side and client side, we should be able to set up a test case where a client authenticate to a server, right?

**Comment by kjellwinblad** [comment]
I noticed that there is another PR https://github.com/kafka4beam/sasl_auth/pull/28 which contains some bug fixes. I think it is better if we can merge that PR first and then do the same type on locking also for the server code (if that is necessary to make the library thread safe).

**Comment by zhongwencool** [comment]
> I noticed that there is another PR #28 which contains some bug fixes. I think it is better if we can merge that PR first and then do the same type on locking also for the server code (if that is necessary to make the library thread safe).

@kjellwinblad  OK，Thank you very much for your review. as I was testing, I realized that no matter how I add the PRINCIPAL of the server, I can't get the kinit to work. can you help me to see if something is not set up correctly?

**Comment by zhongwencool** [comment]
Ping @kjellwinblad @starbelly Do you have time to review this PR?

**Comment by starbelly** [comment]
> Ping @kjellwinblad @starbelly Do you have time to review this PR?

I'm happy to give a review, but it won't be fast 😄 I'll try to circle back to this later in the week.

**Comment by zhongwencool** [comment]
ping @starbelly

File: c_src/sasl_auth.c [hunk 5]
```diff
@@ -129,13 +128,15 @@ static int load(ErlNifEnv* env, void** UNUSED(priv), ERL_NIF_TERM UNUSED(info))
...
         (sasl_mutex_unlock_t*)&sasl_mutex_unlock, (sasl_mutex_free_t*)&sasl_mutex_free);
 
-    result = sasl_client_init(NULL);
-    return !sasl_resource && !(result == SASL_OK);
+    sasl_client_connection_nif_resource_type = init_resource_type(env, "sasl_auth_cli_state");
+    int cli_result = sasl_client_init(NULL);
+    sasl_server_connection_nif_resource_type = init_resource_type(env, "sasl_auth_srv_state");
+    int srv_result = sasl_server_init(NULL, "sasl_auth");
+    return !sasl_client_connection_nif_resource_type && !(cli_result == SASL_OK)
+        && !sasl_server_connection_nif_resource_type && !(srv_result == SASL_OK);
```
Comment (zmstone) [inline c_src/sasl_auth.c:139]: when server state failed to init, it should release client resource before return?

File: scripts/docker_run.sh [hunk 6]
```diff
@@ -1,12 +1,15 @@
 #!/bin/sh
 docker run --rm -it \
+--hostname example.com \
 --name sasl_auth \
 -e SASL_AUTH_TEST_HOST=$SASL_AUTH_TEST_HOST \
 -e SASL_AUTH_TEST_KEY_TAB=$SASL_AUTH_TEST_KEY_TAB \
 -e SASL_AUTH_TEST_PRINCIPAL=$SASL_AUTH_TEST_PRINCIPAL \
+--e SASL_AUTH_SERVICE_HOST=$SASL_AUTH_SERVICE_HOST \
+--e SASL_AUTH_SERVICE_KEY_TAB=$SASL_AUTH_SERVICE_KEY_TAB \
+--e SASL_AUTH_SERVICE_PRINCIPAL=$SASL_AUTH_SERVICE_PRINCIPAL \
```
Comment (zmstone) [inline scripts/docker_run.sh]:
```suggestion
-e SASL_AUTH_SERVICE_HOST=$SASL_AUTH_SERVICE_HOST \
-e SASL_AUTH_SERVICE_KEY_TAB=$SASL_AUTH_SERVICE_KEY_TAB \
-e SASL_AUTH_SERVICE_PRINCIPAL=$SASL_AUTH_SERVICE_PRINCIPAL \
```

File: c_src/sasl_auth.c [hunk 7]
```diff
@@ -41,8 +41,9 @@ typedef struct {
     int mech_set;
 } sasl_state_t;
 
-// SASL connection state
-static ErlNifResourceType* sasl_resource = NULL;
+// SASL connection nif resource
+static ErlNifResourceType* sasl_client_connection_nif_resource_type = NULL;
```
Comment (zmstone) [inline c_src/sasl_auth.c:45]: this naming is a bit odd to me, maybe 
```suggestion
static ErlNifResourceType* sasl_client_connection_nif_resource = NULL;
```

File: c_src/sasl_auth.c (hunk 5 above)
Comment (zmstone) [reply c_src/sasl_auth.c:139]: there was no enif_release_resource for the resource type though
maybe both should be released in unload() ?

Comment (zhongwencool) [reply c_src/sasl_auth.c:139]:
```
static ErlNifResourceType* init_resource_type(ErlNifEnv* env, const char* name)
{
    return enif_open_resource_type(env, NULL, name, (ErlNifResourceDtor*)destroy_resource,
        ERL_NIF_RT_CREATE | ERL_NIF_RT_TAKEOVER, NULL);
}
```
Already passed the `destroy_resource` callback(destructor function) in the enif_open_resource_type. don't need to destroy_resource on unload.

> typedef void ErlNifResourceDtor(ErlNifEnv* caller_env, void* obj);
>
>The function prototype of a resource destructor function.
>The obj argument is a pointer to the resource. The only allowed use for the resource in the destructor is to access its user data one final time. The destructor is guaranteed to be the last callback before the resource is deallocated.

**Review by zmstone — APPROVED** [review]

**Comment by zmstone** [comment]
v2.1.0

**Pull request merged**